File: services/player-service/app/domain/enums.py
from enum import StrEnum


class ValorantRole(StrEnum):
    DUELIST = "DUELIST"
    CONTROLLER = "CONTROLLER"
    INITIATOR = "INITIATOR"
    SENTINEL = "SENTINEL"


class PlayerStatus(StrEnum):
    OFFLINE = "OFFLINE"
    ONLINE = "ONLINE"
    READY_TO_PLAY = "READY_TO_PLAY"
    IN_GAME = "IN_GAME"
    BUSY = "BUSY"


class ValorantRank(StrEnum):
    IRON_1 = "IRON_1"
    IRON_2 = "IRON_2"
    IRON_3 = "IRON_3"
    BRONZE_1 = "BRONZE_1"
    BRONZE_2 = "BRONZE_2"
    BRONZE_3 = "BRONZE_3"
    SILVER_1 = "SILVER_1"
    SILVER_2 = "SILVER_2"
    SILVER_3 = "SILVER_3"
    GOLD_1 = "GOLD_1"
    GOLD_2 = "GOLD_2"
    GOLD_3 = "GOLD_3"
    PLATINUM_1 = "PLATINUM_1"
    PLATINUM_2 = "PLATINUM_2"
    PLATINUM_3 = "PLATINUM_3"
    DIAMOND_1 = "DIAMOND_1"
    DIAMOND_2 = "DIAMOND_2"
    DIAMOND_3 = "DIAMOND_3"
    ASCENDANT_1 = "ASCENDANT_1"
    ASCENDANT_2 = "ASCENDANT_2"
    ASCENDANT_3 = "ASCENDANT_3"
    IMMORTAL_1 = "IMMORTAL_1"
    IMMORTAL_2 = "IMMORTAL_2"
    IMMORTAL_3 = "IMMORTAL_3"
    RADIANT = "RADIANT"

    @property
    def order(self) -> int:
        """Return the rank's position in the hierarchy."""
        return tuple(type(self)).index(self)

    def is_between_or_equal(self, minimum: "ValorantRank", maximum: "ValorantRank") -> bool:
        """
        Check whether this rank belongs to given rank range
        :param minimum: the lowest possible rank
        :param maximum: the highest possible rank
        :return: True if rank belongs to given range False otherwise
        """
        if minimum.order > maximum.order:
            raise ValueError("minimum rank cannot be higher than maximum rank")
        return minimum.order <= self.order <= maximum.order

File: services/matchmaking-service/internal/domain/enums.go
package domain

var valorantRankOrder = map[string]int{
	"IRON_1": 1,
	"IRON_2": 2,
	"IRON_3": 3,

	"BRONZE_1": 4,
	"BRONZE_2": 5,
	"BRONZE_3": 6,

	"SILVER_1": 7,
	"SILVER_2": 8,
	"SILVER_3": 9,

	"GOLD_1": 10,
	"GOLD_2": 11,
	"GOLD_3": 12,

	"PLATINUM_1": 13,
	"PLATINUM_2": 14,
	"PLATINUM_3": 15,

	"DIAMOND_1": 16,
	"DIAMOND_2": 17,
	"DIAMOND_3": 18,

	"ASCENDANT_1": 19,
	"ASCENDANT_2": 20,
	"ASCENDANT_3": 21,

	"IMMORTAL_1": 22,
	"IMMORTAL_2": 23,
	"IMMORTAL_3": 24,

	"RADIANT": 25,
}

var validValorantRoles = map[string]struct{}{
	"DUELIST":    {},
	"CONTROLLER": {},
	"INITIATOR":  {},
	"SENTINEL":   {},
}

var validPlayerStatuses = map[string]struct{}{
	"OFFLINE":       {},
	"ONLINE":        {},
	"READY_TO_PLAY": {},
	"IN_GAME":       {},
	"BUSY":          {},
}

var validMatchingStrategies = map[string]struct{}{
	"BALANCED":       {},
	"RANK_FOCUSED":   {},
	"RATING_FOCUSED": {},
	"ROLE_FOCUSED":   {},
}

func IsValidValorantRank(rank string) bool {
	_, ok := valorantRankOrder[rank]
	return ok
}

func IsValidValorantRole(role string) bool {
	_, ok := validValorantRoles[role]
	return ok
}

func IsValidPlayerStatus(status string) bool {
	_, ok := validPlayerStatuses[status]
	return ok
}

func IsValidMatchingStrategy(strategy string) bool {
	_, ok := validMatchingStrategies[strategy]
	return ok
}

func IsRankRangeValid(minRank string, maxRank string) bool {
	minRankOrder, minRankExists := valorantRankOrder[minRank]
	maxRankOrder, maxRankExists := valorantRankOrder[maxRank]

	if !minRankExists || !maxRankExists {
		return false
	}

	return minRankOrder <= maxRankOrder
}

func IsRankInRange(rank string, minRank string, maxRank string) bool {
	rankOrder, rankExists := valorantRankOrder[rank]
	minRankOrder, minRankExists := valorantRankOrder[minRank]
	maxRankOrder, maxRankExists := valorantRankOrder[maxRank]

	if !rankExists || !minRankExists || !maxRankExists {
		return false
	}

	return rankOrder >= minRankOrder && rankOrder <= maxRankOrder
}
